nav2_behavior_tree/plugins/decorator/goal_updated_controller.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "behaviortree_cpp_v3/decorator_node.h"
+#include "nav2_behavior_tree/plugins/decorator/goal_updated_controller.hpp"
+
+
+namespace nav2_behavior_tree
+{
+
+GoalUpdatedController::GoalUpdatedController(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::DecoratorNode(name, conf),
+first_time_(false)
+{
+}
+
+BT::NodeStatus GoalUpdatedController::tick()
+{
+  if (status() == BT::NodeStatus::IDLE) {
+    // Reset since we're starting a new iteration of
+    // the goal updated controller (moving from IDLE to RUNNING)
+    first_time_ = true;
+    goal_was_updated_ = false;
+  }
+
+  setStatus(BT::NodeStatus::RUNNING);
+
+  if (first_time_) {
```

Comment:
I don't think `first_time_` is required, if you place the 2 initial `get`s inside of the `if IDLE` condition above. I'm not sure (?) that `first_time_` is required in the main if in L65. 

Thinking through the logic, if this is called a 2nd time, the `goal_was_updated_` will be true. If the child action is still continuing, it'll tick again. So its really just the first time we need to specifically say to go for it. 

We could just in `if (status() == BT::NodeStatus::IDLE) {` say `goal_was_updated_ = true` since its updated from the idle state reference. That would be the same logic without the `first_time_` variable.

Reply:
I agree with this, `first_time_` isn't required.

Although, does it make sense for this node to tick its child when it's updated from IDLE state to RUNNING state on the first tick? The node is only meant to tick its child when goal is updated on the blackboard. But with this logic, we assume that the decorator should tick its child on the first tick and any further ticks where the goal is updated. This logic made sense for `RateController` and `DistanceController` where we start the logic at 0 seconds or 0 meters traveled but I'm not sure if `GoalUpdated` condition requires a mandatory execution of its child on its first tick, that should happen automatically on the first tick when `if (goal_ != current_goal || goals_ != current_goals)` condition is true.

We could just remove `first_time_` and place to 2 initial `get`s inside the `if (status() == BT::NodeStatus::IDLE)` condition while setting `goal_was_updated_ = false` within the condition and leave everything else as it is. That would mean that this node would tick its child only when the goal is updated on the blackboard.

Reply:
1) Fully agree with the `first_time` parameter, I wanted to keep it to make it obvious but since the code is so easy, it's quickly figured out. 

2) I think from the naming it makes sense not to tick the child on the first visit, I am just wondering if application-wise it makes sense to tick it anyway for the first time. Either way, it's a quick change.

Reply:
We should always tick on the first visit for this kind of thing - or else when you submitted a navigation request, it would never plan at all (you'd have to issue another new goal afterwards)